### PR TITLE
Added Summary Table to Simulation Page

### PR DESCRIPTION
**Pull request by jennygaoo**
The table is still very basic. I've searched around, and it seems like the Google Charts API cannot create nested rows. Will continue to see if it's possible to do this. The table still needs sorting, pagination, color coding cells (the table doesn't seem user friendly). 
Here's what it looks like as of now!
Screenshot 2020-07-21 at 10 27 39 AM

### CODE REVIEW

**Comment by jennygaoo** [comment]
So this branch, "populate-table", is a child branch of master, not "user-input-form-validation". Ideally, it would've been. 

I started working on "populate-table" prior to making the other branch. However, I made a PR for "user-input-form-validation" before "populate-table". Since "populate-table" is not a child branch of "user-input-form-validation", the commits from "user-input-form-validation" show up in this PR.

File: postal-injection/src/main/webapp/summary_table.js [hunk 1]
```diff
@@ -5,12 +5,71 @@ async function createSummaryTable() {
...
+      const currentPostalInjectionShippingCost =
+          summaryTableData[row][subRow].postalInjectionCost;
+      const currentNumberOfOrders =
+          summaryTableData[row][subRow].numberOfOrders;
+
+      /* if the cost difference or postal injection shipping cost doesn't apply,
+      display "unavailable" */
+      data.addRow([
+        currentZipCode, currentWarehouse,
+        currentCostDifference === -1 ? {v: -1, f: 'unavailable'} : {
```
Comment (jennygaoo) [inline postal-injection/src/main/webapp/summary_table.js:45]: v: value, f: what's actually displayed. Added this as a comment in the code (if you're viewing the "added data display table" commit, you can't see this comment)

File: postal-injection/src/main/webapp/summary_table.js [hunk 2]
```diff
@@ -2,11 +2,155 @@
...
+          summaryTableData[row][subRow].postalInjectionCost;
+      const currentNumberOfOrders =
+          summaryTableData[row][subRow].numberOfOrders;
+
+      /* if the cost difference or postal injection shipping cost doesn't apply,
+      display "unavailable" */
+      data.addRow([
+        currentZipCode, currentWarehouse,
+        // v: value, f: what's actually displayed
+        currentCostDifference === -1 ? {v: -1, f: 'unavailable'} : {
```
Comment (jennygaoo) [inline postal-injection/src/main/webapp/summary_table.js:45]: When it's not possible to do "cost difference" and "postal injection", I've set the value of the cell to -1. I don't think this is the best solution because if I were to sort the values, I'd get "build-ups" of these values...

The "cost difference" can be both negative and positive, so if I were to sort the cost differences by value, I'd get a bunch of these "unavailable" rows between the negative and positive values. I will look into whether or not I can change the value to negative infinity or something like that.

File: postal-injection/src/main/webapp/summary_table.js [hunk 3]
```diff
@@ -2,11 +2,155 @@
...
+  data.addColumn('string', 'Warehouse');
+  data.addColumn(
+      'number',
+      'Cost Difference Between Postal Injection and Regular Shipping');
+  data.addColumn('number', 'Regular Shipping Cost');
+  data.addColumn('number', 'Postal Injection Cost');
+  data.addColumn('number', 'Number of Orders');
+
+  for (const row in summaryTableData) {
+    for (const subRow in summaryTableData[row]) {
```
Comment (jennygaoo) [inline postal-injection/src/main/webapp/summary_table.js:28]: the nested for loop looks like a bottleneck (esp for all the zip codes). However, it only takes a few seconds to create the table for all of the zip codes. Will look into optimizing this though

File: postal-injection/src/main/webapp/summary_table.js [hunk 4]
```diff
@@ -2,11 +2,155 @@
...
+
+  if (!hasValidThreshold) {
+    alert(
+        'Please ensure that the order threshold is a positive integer greater than or equal to 1.');
+    return false;
+  }
+  return true;
 }
 
 function getUserInputs() {
```
Comment (k-sharpe) [inline postal-injection/src/main/webapp/summary_table.js:156]: Kinda think this should return an object not a bunch of assorted values

File: postal-injection/src/main/webapp/summary_table.js [hunk 5]
```diff
@@ -2,11 +2,155 @@
 /* eslint-disable */
 
 async function createSummaryTable() {
-  const [zipCode, regularRate, palletRate, minPalletSize, threshold, postOfficeToCustomerRate] =
-      getUserInputs();
-  fetchTableData(
-      zipCode, regularRate, palletRate, minPalletSize, threshold,
-      postOfficeToCustomerRate);
+  if (checkRequiredInputs() && checkCorrectInputTypes()) {
+    const [zipCode, regularRate, palletRate, minPalletSize, threshold, postOfficeToCustomerRate] =
```
Comment (k-sharpe) [inline postal-injection/src/main/webapp/summary_table.js:6]: Yea i think getting an object here might be more readable because I was thinking maybe someone could change around the getUserInputs function and then everything might break

File: postal-injection/src/main/webapp/summary_table.js (hunk 1 above)
Comment (k-sharpe) [reply postal-injection/src/main/webapp/summary_table.js:45]: is it not possible to change the variables to reflect what they are? the comment helps a lot though

File: postal-injection/src/main/webapp/summary_table.js [hunk 6]
```diff
@@ -2,11 +2,155 @@
...
+function checkCorrectInputTypes() {
+  return checkZipCodeInput() && checkShippingRateInputs() &&
+      checkMinPalletSizeInput() && checkThresholdInput();
+}
+
+/** checks that the zip code input is a positive integer with only 5 digits */
+function checkZipCodeInput() {
+  // cast zip-code input as Number to check if it's an int
+  const zipCode = document.getElementById('zip-code').value
+  const zipCodeAsNumber = Number(zipCode);
```
Comment (k-sharpe) [inline postal-injection/src/main/webapp/summary_table.js:95]: This will not work, as some zip codes have the additional sub code for delivery EX: 95135-2207 and these are in our data set

File: postal-injection/src/main/webapp/summary_table.js [hunk 7]
```diff
@@ -2,11 +2,155 @@
...
+  /* zip code input is either blank (returns all zip codes),
+  or a positive 5 digit integer zip code */
+  const hasValidZipcode = (zipCode === '') ||
+      ((Number.isInteger(zipCodeAsNumber) && (zipCode.length === 5) &&
```
Comment (k-sharpe) [inline postal-injection/src/main/webapp/summary_table.js:99]: yea see above, zip codes are up to 9 digits as a string "95101-2023"

File: postal-injection/src/main/webapp/summary_table.js (hunk 1 above)
Comment (jfewii) [reply postal-injection/src/main/webapp/summary_table.js:45]: +1

File: postal-injection/src/main/webapp/summary_table.js [hunk 8]
```diff
@@ -2,11 +2,155 @@
...
+       (postOfficeToCustomerRateAsNumber > 0));
+  if (!hasValidShippingRates) {
+    alert('Please ensure shipping rates are positive.');
+    return false;
+  }
+  return true
+}
+
+/** checks that the minimim pallet size is a positive integer */
+function checkMinPalletSizeInput() {
```
Comment (jfewii) [inline postal-injection/src/main/webapp/summary_table.js:128]: Is there a way I can call your check input functions, or do I have to use your same code?

File: postal-injection/src/main/webapp/summary_table.js (hunk 1 above)
Comment (andreahu12) [reply postal-injection/src/main/webapp/summary_table.js:45]: Agreed. If possible, it would be nice for v and f to be more descriptive. If not, you can also include that detail in your comment since comments are here to explain _why_ :)

nit: functionally identical but please change to `{v: currentCostDifference, f: 'unavailable'}` to be more similar to the other ones/ less error prone.

File: postal-injection/src/main/webapp/summary_table.js [hunk 9]
```diff
@@ -2,11 +2,155 @@
...
+  const table =
+      new google.visualization.Table(document.getElementById('data_table'));
+  table.draw(
+      data,
+      {allowHtml: true, showRowNumber: true, width: '100%', height: '100%'});
+}
+
+function checkRequiredInputs() {
+  // get all the inputs within the submitted form
+  var inputs = document.getElementsByTagName('input');
```
Comment (andreahu12) [inline postal-injection/src/main/webapp/summary_table.js:76]: use const instead of var

File: postal-injection/src/main/webapp/summary_table.js (hunk 6 above)
Comment (andreahu12) [reply postal-injection/src/main/webapp/summary_table.js:95]: I just scrolled through allOrdersWithDistances and the zipcodes all look like the first 5 digits to me. However I think in some cases, zipcodes start with 0 so something like 02120 gets truncated to 2120. We should double check those are handled, but perhaps in a followup.

File: postal-injection/src/main/webapp/summary_table.js (hunk 8 above)
Comment (jennygaoo) [reply postal-injection/src/main/webapp/summary_table.js:128]: I think there's a way for you to call it
https://stackoverflow.com/questions/25962958/calling-a-javascript-function-in-another-js-file

File: postal-injection/src/main/webapp/summary_table.js [hunk 10]
```diff
@@ -18,20 +166,25 @@ function getUserInputs() {
...
+  const userInputObject = {
+    zipCode: zipCode,
+    regularRate: regularRate,
+    palletRate: palletRate,
+    minPalletSize: minPalletSize,
+    threshold: threshold,
+    postOfficeToCustomerRate: postOfficeToCustomerRate
+  } 
+  return userInputObject;
 }
```
Comment (jennygaoo) [inline postal-injection/src/main/webapp/summary_table.js:178]: so "make pretty" puts "return userInputObject()" on line 176 after the }, but "make validate" says this formatting is an error. Will look at how I can "override"/dismiss the "make pretty" here

File: postal-injection/src/main/webapp/summary_table.js (hunk 3 above)
Comment (andreahu12) [reply postal-injection/src/main/webapp/summary_table.js:29]: can add // TODO(jennyagao): optimize calculation

File: postal-injection/src/main/webapp/summary_table.js (hunk 10 above)
Comment (jennygaoo) [reply postal-injection/src/main/webapp/summary_table.js:178]: fixed this by creating the JS object another way!

**Comment by jgao23** [comment]
> **TL; DR:** check all commits from "added data display table" and onwards to see the changes that are actually relevant to this PR.
> 
> So this branch, "populate-table", is a child branch of master, not "user-input-form-validation". Ideally, it would've been.
> 
> I started working on "populate-table" prior to making the "user-input-form-validation".
> 
> However, I made a PR for "user-input-form-validation" before "populate-table". Since "populate-table" is not a child branch of "user-input-form-validation", the commits from "user-input-form-validation" show up in this PR.

what happens if you set the parent branch to the user-input-form-validation branch? will we still see the changes from user-input-form-validation?

File: postal-injection/src/main/webapp/summary_table.js (hunk 8 above)
Comment (jgao23) [reply postal-injection/src/main/webapp/summary_table.js:133]: if you are planning on sharing this code, you may want to create a new file that holds common functions/constants :)

File: postal-injection/src/main/webapp/summary_table.js [hunk 11]
```diff
@@ -2,11 +2,160 @@
...
+              f: '$' + currentCostDifference.toString()
+            },
+        +currentRegularShippingCost,
+        +currentPostalInjectionShippingCost === -1 ?
+            {v: currentPostalInjectionShippingCost, f: 'unavailable'} :
+            {
+              v: currentPostalInjectionShippingCost,
+              f: '$' + currentPostalInjectionShippingCost.toString()
+            },
+        +currentNumberOfOrders
```
Comment (jgao23) [inline postal-injection/src/main/webapp/summary_table.js:61]: nit: maybe it's just me being dumb with js, but this is unreadable to me. i think what you're doing is adding a row of values, and you're doing the calculation within the array of values itself? instead of doing that, i think it would be more readable if break out the calculation and assign the value to variables.

File: postal-injection/src/main/webapp/summary_table.js [hunk 12]
```diff
@@ -2,11 +2,160 @@
...
+    loaded */
+    google.charts.setOnLoadCallback(function() {
+      drawZipCodeSummaryTable(summaryTableData)
+    });
+  }
+}
+
+function drawZipCodeSummaryTable(summaryTableData) {
+  const data = new google.visualization.DataTable();
+  data.addColumn('string', 'Zip Code');
```
Comment (jgao23) [inline postal-injection/src/main/webapp/summary_table.js:19]: supernit: add new empty line, then add a comment that you're adding the columns to the table

File: postal-injection/src/main/webapp/summary_table.js [hunk 13]
```diff
@@ -2,11 +2,160 @@
...
+  data.addColumn('string', 'Warehouse');
+  data.addColumn(
+      'number',
+      'Cost Difference Between Postal Injection and Regular Shipping');
+  data.addColumn('number', 'Regular Shipping Cost');
+  data.addColumn('number', 'Postal Injection Cost');
+  data.addColumn('number', 'Number of Orders');
+
+  // TODO(jennyagao): optimize calculation
+  for (const row in summaryTableData) {
```
Comment (jgao23) [inline postal-injection/src/main/webapp/summary_table.js:29]: supernit: add a comment that you're adding all the rows to the table

File: postal-injection/src/main/webapp/summary_table.js [hunk 14]
```diff
@@ -2,11 +2,160 @@
...
+              f: '$' + currentPostalInjectionShippingCost.toString()
+            },
+        +currentNumberOfOrders
+      ]);
+    }
+  }
+
+  /* can't apply formatting to current cost difference
+  & postal injection columns because not all of them are numbers */
+  const formatter = new google.visualization.NumberFormat({prefix: '$'});
```
Comment (jgao23) [inline postal-injection/src/main/webapp/summary_table.js:68]: supernit: amend comment to something like:
`/* Formatting the table. Can't apply number formatting to cost difference and postal injection columns because they could be 'unavailable'. */

File: postal-injection/src/main/webapp/summary_table.js [hunk 15]
```diff
@@ -2,11 +2,160 @@
...
+  // apply formatter to regular shipping cost column
+  formatter.format(data, 3);
+
+  const table =
```
Comment (jgao23) [inline postal-injection/src/main/webapp/summary_table.js:72]: supernit: add comment that the below code draws the table in the element
